pxf/pxf-service/src/test/java/org/apache/hawq/pxf/service/UGICacheTest.java

```diff
@@ -0,0 +1,197 @@
+        UGICacheEntry proxyUGI1 = cache.getTimedProxyUGI(session);
+
+        cache.release(proxyUGI1, false);
+        proxyUGI1.setExpired();
+        verify(provider, never()).destroy(any(UserGroupInformation.class));
+        UGICacheEntry proxyUGI2 = cache.getTimedProxyUGI(session);
+        verify(provider).destroy(proxyUGI1.getUGI());
+        assertNotEquals(proxyUGI2, proxyUGI1);
+        assertNotEquals(proxyUGI2.getUGI(), proxyUGI1.getUGI());
+        verify(provider, never()).destroy(proxyUGI2.getUGI());
+        // this does not clean up any UGIs because our proxyUGI is still in use.
+        assertNotEquals(proxyUGI2, proxyUGI1);
+        verify(provider, times(2)).createProxyUGI("the-user");
+    }
+
+    @Test
+    public void releaseAnExpiredUGIResetsTheTimer() throws Exception {
+        UGICacheEntry proxyUGI1 = cache.getTimedProxyUGI(session);
+        proxyUGI1.setExpired();
+        cache.release(proxyUGI1, false);
+        UGICacheEntry proxyUGI2 = cache.getTimedProxyUGI(session);
+        assertEquals(proxyUGI2, proxyUGI1);
+        verify(provider, never()).destroy(any(UserGroupInformation.class));
+    }
+
+    @Test
+    public void releaseDoesNotFreeResourcesIfUGIIsUsedElsewhere() throws Exception {
+        UGICacheEntry proxyUGI1 = cache.getTimedProxyUGI(session);
+        UGICacheEntry proxyUGI2 = cache.getTimedProxyUGI(session);
+
+        cache.release(proxyUGI1, true);
+        // UGI was not cleaned up because proxyUGI2 is referencing it
+        UGICacheEntry proxyUGI3 = cache.getTimedProxyUGI(session);
+        assertEquals(proxyUGI1, proxyUGI2);
+        assertEquals(proxyUGI1, proxyUGI3);
+        verify(provider, times(1)).createProxyUGI("the-user");
+    }
+
+    @Test
+    public void releasingAllReferencesFreesResources() throws Exception {
+        UGICacheEntry proxyUGI1 = cache.getTimedProxyUGI(session);
+        UGICacheEntry proxyUGI2 = cache.getTimedProxyUGI(session);
+
+        cache.release(proxyUGI1, true);
+        cache.release(proxyUGI2, true);
+        // at this point, the initial UGI has been freed. Calling
+        // getTimedProxyUGI again creates a new UGI.
+        UGICacheEntry proxyUGI3 = cache.getTimedProxyUGI(session);
+        assertEquals(proxyUGI1, proxyUGI2);
+        assertNotEquals(proxyUGI1, proxyUGI3);
+        verify(provider, times(2)).createProxyUGI("the-user");
+    }
+
+    @Test
+    public void releaseAnEntryNotInTheCache() throws Exception {
+        // this could happen if some caller of the cache
+        // retains a reference to a cache entry after releasing it.
+        UGICacheEntry proxyUGI1 = cache.getTimedProxyUGI(session);
+
+        cache.release(proxyUGI1, true);
+        cache.release(proxyUGI1, true);
+        UGICacheEntry proxyUGI2 = cache.getTimedProxyUGI(session);
+        assertNotEquals(proxyUGI1, proxyUGI2);
+        verify(provider, times(2)).createProxyUGI("the-user");
+    }
+}
```

Comment:
good set of tests, I wonder if we need to include tests that use real timer (instead of explicitly setting expired) to make sure "clean-on-get" works and if we need multithreaded testing to make sure there are no deadlocks and sync is correct.

Reply:
To the first point—we can probably mock the system time instead of using real time, but I agree. Multithreaded tests would be good.